Objects/longobject.c

```diff
@@ -4491,38 +4494,65 @@ long_rshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
 {
     PyLongObject *z = NULL;
     Py_ssize_t newsize, hishift, i, j;
-    digit lomask, himask;
+    digit lomask, himask, omitmark;
 
-    if (Py_SIZE(a) < 0) {
-        /* Right shifting negative numbers is harder */
-        PyLongObject *a1, *a2;
-        a1 = (PyLongObject *) long_invert(a);
-        if (a1 == NULL)
-            return NULL;
-        a2 = (PyLongObject *) long_rshift1(a1, wordshift, remshift);
-        Py_DECREF(a1);
-        if (a2 == NULL)
-            return NULL;
-        z = (PyLongObject *) long_invert(a2);
-        Py_DECREF(a2);
+    /* for a positive integrate x
+         -x >> m = -(x >> m)      when the dropped bits are all zeros,
+                 = -(x >> m) -1   otherwise. */
+
+    if (IS_MEDIUM_VALUE(a)) {
+        stwodigits sval;
+        if (wordshift > 0) {
+            return get_small_int(-(Py_SIZE(a) < 0));
+        }
+        sval = Py_ARITHMETIC_RIGHT_SHIFT(stwodigits, medium_value(a), remshift);
+        if (IS_SMALL_INT(sval)) {
+            return get_small_int((sdigit)sval);
+        }
+        z = _PyLong_New(Py_SIZE(a));
+        z->ob_digit[0] = (digit)(a->ob_digit[0] >> remshift);
+        if (Py_SIZE(a) < 0 && (a->ob_digit[0] & ~(PyLong_MASK << remshift))) {
+            z->ob_digit[0] += 1;
+        }
+        return (PyObject *)z;
     }
-    else {
-        newsize = Py_SIZE(a) - wordshift;
-        if (newsize <= 0)
-            return PyLong_FromLong(0);
-        hishift = PyLong_SHIFT - remshift;
-        lomask = ((digit)1 << hishift) - 1;
-        himask = PyLong_MASK ^ lomask;
-        z = _PyLong_New(newsize);
-        if (z == NULL)
-            return NULL;
-        for (i = 0, j = wordshift; i < newsize; i++, j++) {
-            z->ob_digit[i] = (a->ob_digit[j] >> remshift) & lomask;
-            if (i+1 < newsize)
-                z->ob_digit[i] |= (a->ob_digit[j+1] << hishift) & himask;
+
+    newsize = Py_ABS(Py_SIZE(a)) - wordshift;
```

Comment:
This may not be big enough in the negative case. Consider the case of `-(2**60 - 1) >> 30` (assuming 30-bit digits). Then `Py_SIZE(a)` is `2`, `wordshift` is `1`, but we need two digits for the result.

Reply:
`z = _PyLong_New(newsize)` is a helper long object, whose value is assigned as `(2**60 - 1) >> 30` instead of `-(2**60 - 1) >> 30`. The `long_sub()` below gives the final result.

Reply:
Ah, okay; thanks.

I'd propose dropping this set of changes from the PR and focusing on the small int case; we can re-introduce these changes later in a separate PR. I think they'll require their own round of iterations, and I'd rather get the small int improvements in quickly.